src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/NativeMemoryTests.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Xunit;
+
+namespace System.Runtime.InteropServices.Tests
+{
+    public unsafe class NativeMemoryTests
+    {
+        [Fact]
+        public void AlignedAllocTest()
+        {
+            void* ptr = NativeMemory.AlignedAlloc(1, 1);
+            Assert.True(ptr != null);
+            NativeMemory.AlignedFree(ptr);
+        }
+
+        [Fact]
+        public void AlignedAllocOOMTest()
+        {
+            Assert.Throws<OutOfMemoryException>(() => NativeMemory.AlignedAlloc(1, nuint.MaxValue));
+        }
+
+        [Fact]
+        public void AlignedAllocZeroAlignmentTest()
+        {
+            Assert.Throws<OutOfMemoryException>(() => NativeMemory.AlignedAlloc(0, 0));
+        }
+
+        [Fact]
+        public void AlignedAllocNonPowerOfTwoAlignmentTest()
+        {
+            Assert.Throws<OutOfMemoryException>(() => NativeMemory.AlignedAlloc(3, 0));
+        }
+
+        [Fact]
+        public void AlignedAllocNonMultipleOfAlignmentTest()
+        {
+            Assert.Throws<OutOfMemoryException>(() => NativeMemory.AlignedAlloc(2, 1));
```

Comment:
Would be it better to throw ArgumentException for these cases?

Reply:
Yes and I've updated locally. 

Also noting that this will throw `ArgumentException` for `Alignment == 0` || `ByteCount == 0`. This is because `aligned_alloc` and `posix_memalign` require `size` to be a multiple of `alignment` and `posix_memalign` requires `alignment` to be a power of two at least as big as `void*`.

I don't think that implicitly rounding `alignment` nor doing `byteCount = Math.Max(byteCount, alignment)` is intuitive behavior here.

Reply:
> I don't think that implicitly rounding `alignment` nor doing `byteCount = Math.Max(byteCount, alignment)` is intuitive behavior here.

In general, I think it is better to make more cases just work instead of forcing all callers to worry about them, having them enumerated in the documentation, etc.

Requiring the alignment to be power of two makes sense. The rest of the restictions - like requiring alignment to be at least `sizeof(void*)` or requiring the size to be multiple of the aligment looks arbitrary to me. It feels like somebody took a bunch of shortcuts and we are stuck with them till today.

Reply:
Many of these are also things that are fairly trivial for the user to address if they don't want to think about it, such as by calling `BitOperations.RoundUpToPowerOf2`
However, doing it in the `NativeMemory.*` APIs adds fixed overhead to these calls, for something that is likely only going to be used by power/advanced users (many of whom may already be doing their own math/fixups to ensure correctness).

I'm already not a huge fan of this being more than just "call the CRT API and return the exact result given" and forcing more expensive logic for edge case scenarios that power users can likely deal with seems undesirable to me, given the scenario.
-- These are getting less and less like "NativeMemory" APIs and more like general .NET "UnsafeMemory" APIs and I'd expect the power users that this type of API will actually benefit would prefer the former, since you can build the latter on top of it.

Reply:
Would the overhead of making these cases just work be more than throwing the exception for them? I won't be susprised it is actually less.

Reply:
The exception case is exceptional. I wouldn't expect users to `try { alloc(size) } catch { alloc(sizeButFixedToBeValid) }` 😄
So yes, this is going to introduce additional overhead for the normal use-case for these methods.

These methods deal with unsafe code and managing native memory, so they are already designed for advanced/power users.
I would expect such users are likely already doing their own size calculations, adjustments, and validation (often for performance reasons).

There is likely a subset of users that will try to use these APIs regardless, but they should hit the limitations of the APIs via docs and exceptions thrown fairly quickly (and this is a good use case for an analyzer if its problematic) and will ultimately be able to fix their code to do the right thing.

Otherwise, I think we aren't addressing what the primary users of this API likely want/need and are instead adding overhead, rather than providing a primitive building block upon which users can themselves build the appropriate helper APIs.

Reply:
We had similar debate in https://github.com/dotnet/runtime/pull/53669/files#r646737136 . 

Both NativeMemory and System.IO.RandomAccess are for advanced users, and both of these discussions are about whether it is better to do this:

```csharp
if (somewhat arbitrary limitation of underlying C/OS API)
    throw exception;
```

or this:

```csharp
if (somewhat arbitrary limitation of underlying C/OS API)
   make it just work
```

I am in the make it just work camp, for the same reasons @stephentoub mentioned. One of the values of .NET has always been in providing easier to use abstractions, with less sharp edges, e.g. by hiding arbitrary limitation of the underlying C/OS APIs.

Reply:
cc: @adamsitnik

Reply:
-- Just noting that I am going to update as you suggested here, but that I don't think its the right thing given what we reviewed and who I believe the target audience is, as detailed below.

> Both NativeMemory and System.IO.RandomAccess are for advanced users

I think there are a number of differences that make these non-comparable such as living in System.Runtime.InteropServices and being an API exposing what is arguably one of the most heavily optimized, profiled, and used native APIs.

> and both of these discussions are about whether it is better to do this:

I don't think this is quite true either. I am very much in the camp that not checking anything and effectively just being a direct shim to the native API is the right thing. If you call the API, you get all the implementation defined behavior and may need to consider it or write your own wrapper (or use a separate wrapper we exposed, akin to the one being discussed here). I'm not even convinced that checking for `nullptr` and throwing `OOM` is the right thing here.

We reviewed this as basically `malloc` and friends and the more it deviates from being "just malloc", the less useful I think it is for users who want "just malloc" and the more it becomes a "general .NET allocation API" which are two entirely different things.

------------------

I'd expect users fall into one of two camps where they either just want to allocate native memory and don't care about the perf and are calling `AllocHGlobal`/friends or they want to allocate native memory and care about perf, in which case they are likely dealing with calling `malloc`/friends themselves. An API that does all of the checks discussed above falls into the former category and I'd largely expect such users aren't going to switch over without some hint from an analyzer. However, such an API likely excludes a number of users in the latter category as it adds potentially undesirable overhead.

Having these checks on `AlignedAlloc` for example adds ~4ms per million `aligned_alloc` calls. This overhead was fairly stable whether measuring small 16-byte allocations or larger 4kb allocations (the CRT implementation profiled on my box, for either size, took about 33.5ms for a direct call and about 37.6ms for our wrapper). If you were wanting to use these APIs to minimize GC overhead for say a game where you are wanting to target 60 (16ms frames) or even 144fps+ (7ms frames) then this overhead rapidly adds up. Adding aggressive inlining largely helps here for constant inputs, but can regress perf for non-constant inputs depending on the callsite.

Reply:
> I'd expect users fall into one of two camps where they either just want to allocate native memory and don't care about the perf and are calling AllocHGlobal/friends or they want to allocate native memory and care about perf, in which case they are likely dealing with calling malloc/friends themselves

As someone looking forward to using these new APIs, I'll just say that I would fall into a third camp. I need more control than `AllocHGlobal` gives -- like allocating aligned memory -- but won't be making allocations so frequently that it's worth learning all the platform-specific details around malloc behavior to save a few nanoseconds.  I generally expect the BCL to follow the "pit of success" model where possible, and here it's *relatively* cheap to do so.

If people fall into the perf-over-usability camp, there's always lower-level options like your mimalloc port.  😄

Reply:
> exposing what is arguably one of the most heavily optimized, profiled, and used native APIs.

I do not agree that aligned alloc is the most heavily optimized, profiled, and used native API.

> Having these checks on AlignedAlloc for example adds ~4ms per million aligned_alloc calls.

Ok, and what is the cost of one million of `aligned_alloc/free` calls plus actually doing something semi-useful with the allocated memory?

This overhead sounds perfectly fine to me - I expect it to be in the noise range.

Reply:
> I do not agree that aligned alloc is the most heavily optimized, profiled, and used native API.

I was referring to the allocation APIs in general with regards to this statement.

> This overhead sounds perfectly fine to me - I expect it to be in the noise range.

That's not my expectation, particularly for many small allocations in a hot path where you may have already done the relevant checks (which is something `malloc` and friends also get used and optimized for).

Something like `mimalloc`, for example, has an implementation that is 2-4x faster than the built-in CRT/Windows implementations and is used by Bing/Azure (https://github.com/microsoft/mimalloc#usage). These additional checks would represent a 25-50% additional overhead per call (8ms for 1m aligned allocations) for a comparative scale.